### PR TITLE
Cleanup BlobPath Class

### PR DESCRIPTION
**Pull request by original-brownbear**
There should be a singleton for the empty version of this.
All the copying to `String[]` or use as an iterator make
no sense either when we can just use the list outright.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobStore.java [hunk 1]
```diff
@@ -70,14 +71,14 @@ private synchronized Path buildAndCreate(BlobPath path) throws IOException {
...
+        List<String> paths = path.parts();
+        if (paths.isEmpty()) {
             return path();
         }
-        Path blobPath = this.path.resolve(paths[0]);
-        if (paths.length > 1) {
-            for (int i = 1; i < paths.length; i++) {
-                blobPath = blobPath.resolve(paths[i]);
+        Path blobPath = this.path.resolve(paths.get(0));
+        if (paths.size() > 1) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobStore.java:79]: This `if` isn't necessary, we check the condition in the `for` loop anyway.

File: modules/repository-url/src/main/java/org/elasticsearch/common/blobstore/url/URLBlobStore.java [hunk 2]
```diff
@@ -112,14 +113,14 @@ public void close() {
...
+        List<String> paths = path.parts();
+        if (paths.isEmpty()) {
             return path();
         }
-        URL blobPath = new URL(this.path, paths[0] + "/");
-        if (paths.length > 1) {
-            for (int i = 1; i < paths.length; i++) {
-                blobPath = new URL(blobPath, paths[i] + "/");
+        URL blobPath = new URL(this.path, paths.get(0) + "/");
+        if (paths.size() > 1) {
```
Comment (DaveCTurner) [inline modules/repository-url/src/main/java/org/elasticsearch/common/blobstore/url/URLBlobStore.java:121]: This `if` isn't necessary, we check the condition in the loop anyway.

File: server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java [hunk 3]
```diff
@@ -12,38 +12,26 @@
...
-        return paths.iterator();
-    }
-
-    public String[] toArray() {
-        return paths.toArray(new String[paths.size()]);
+    public List<String> parts() {
+        return paths;
     }
 
     public BlobPath add(String path) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java:37]: This method uses `CollectionUtils#appendToCopy` which returns a mutable list. Either we should make it unmodifiable here, or else change that method to make it always return something unmodifiable.

I believe the other two constructor calls give us an unmodifiable collection already, it's just this one place.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a couple of nits.

File: modules/repository-url/src/main/java/org/elasticsearch/common/blobstore/url/URLBlobStore.java (hunk 2 above)
Comment (original-brownbear) [reply modules/repository-url/src/main/java/org/elasticsearch/common/blobstore/url/URLBlobStore.java]: ++

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobStore.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobStore.java]: ++

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**